source/class.cpp

```diff
@@ -737,9 +737,22 @@ string bind_member_functions_for_call_back(CXXRecordDecl const *C, string const
 
 				c += "\t{} {}({}){} {}override {{"_format(return_type, m->getNameAsString(), std::get<0>(args), m->isConst() ? " const" : "", exception_specification);
 
-				c += indent(fmt::format(call_back_function_body_template, class_name, /*class_qualified_name(C), */ python_name, std::get<1>(args), return_type), "\t\t");
-				if( m->isPure() ) c += "\t\tpybind11::pybind11_fail(\"Tried to call pure virtual function \\\"{}::{}\\\"\");\n"_format(C->getNameAsString(), python_name);
-				else c += "\t\treturn {}::{}({});\n"_format(C->getNameAsString(), m->getNameAsString(), std::get<1>(args));
+				string member_function_name = namespace_from_named_decl(C);
+				if ( member_function_name.length() > 0 ) member_function_name += "::";
+				member_function_name +=  C->getNameAsString() + "::" + m->getNameAsString();
```

Comment:
I am confused, for the test T80 here we need `member_function_name` to be `aaa::A::foo` (so that we don't have to write one line for every template type) and `function_qualified_name` returns `aaa::A<double>::foo(int)` and `class_name` is `aaa::A<double>`.

Do you recommend a different way to obtain  `aaa::A::foo`?

Reply:
- good point! Let's see... since we might need just-the-function name in trampoline code (ie like here we need string "Eat" https://github.com/RosettaCommons/binder/blob/master/test/T10.virtual_inheritance.ref#L27) then i think we should pass _just_ the member function name (without even the class name, `m->getNameAsString()`). (My thoughts is to allow custom code to fully emulate Binder default behavior if needed).

- as for class name, i am leaning towards passing `aaa::A<double>` (type name + namespace + template specialization) so we can print full class info when needed (ie error message). Thoughts?

Reply:
I think that there might be a miss-understanding here. 
The `member_function_name` is only used to check the configuration option. This does not appear on the generated files. The role of `member_function_name` is to be used as a key to access a `std::map` to know if the user provides a user-defined function and (if yes) what is the name of the user defined function. I would be in favor not to include template name in the configuration script for this as it would require modification of the script to include every template instances.

As an example: 
The line https://github.com/kliegeois/binder/blob/trampoline_custom/test/T80.custom_trampoline.config#L2 will generate the code: https://github.com/kliegeois/binder/blob/trampoline_custom/test/T80.custom_trampoline.ref#L22 .
Note that the line https://github.com/RosettaCommons/binder/blob/master/test/T10.virtual_inheritance.ref#L27 could be fully emulated from within `myFoo` call as `myFoo` has access to the 2 template types (`A` and its trampoline) and the function name as a string argument.

Currently, the code does not pass the class name but pass the class as a template argument as the user-defined function might need to create instance of the class or cast to a pointer of that class. The function name is important to pass though to be able to reproduce https://github.com/RosettaCommons/binder/blob/master/test/T10.virtual_inheritance.ref#L27 .

Reply:
ah, i see, - sorry i misread this! Ok, in that case we can merge this part as-is. In previous cases i did check for both template and non-template functions in config arguments (having both of these checks proven to be useful in my project). If you would like to implement similar functionality (optional) please see https://github.com/RosettaCommons/binder/blob/master/source/function.cpp#L326

> Currently, the code does not pass the class name but pass the class as a template argument as the user-defined

-- it might be good to pass class name as-string as well. So if needed we can generate more informative error message (see https://github.com/RosettaCommons/binder/blob/master/test/T10.virtual_inheritance.ref#L36) which we unfortunately can not do from template parameter. Imagine the situation when error happened in runtime in function named `clone` - error message from without class name would be of little use since name is so common.